content/academy/puppeteer_playwright/collecting_data.md

```diff
@@ -0,0 +1,284 @@
+---
+title: IV - Collecting data
+description: Learn how to collect data from a page with evaluate functions, then how to safely collect it by using a second library called Cheerio.
+menuWeight: 7.4
+paths:
+    - puppeteer-playwright/collecting-data
+---
+
+# [](#collecting-data) Collecting data
+
+Now that we know how to execute scripts on a page, we're ready to learn a bit about [data collection]({{@link web_scraping_for_beginners/data_collection.md}}). In this lesson, we'll be scraping all of the on-sale products from our [Fakestore](https://demo-webstore.apify.org/search/on-sale) website.
+
+> Most data collection cases involve looping through a list of items of some sort.
+
+There are two main ways to collect data with Playwright and Puppeteer:
+
+1. Directly in `page.evaluate()` and other evaluate functions such as `page.$$eval()`.
+2. In the Node.js context using a parsing library such as [Cheerio](https://www.npmjs.com/package/cheerio)
+
+<!-- Understanding the two different contexts your code is running is  -->
+
+## Setup
+
+Here is the base set up for our code, upon which we'll be building off of in this lesson:
+
+```marked-tabs
+<marked-tab header="Playwright" lang="javascript">
+import { chromium } from 'playwright';
+
+const browser = await chromium.launch({ headless: false });
+const page = await browser.newPage();
+
+await page.goto('https://demo-webstore.apify.org/search/on-sale');
+
+await page.waitForLoadState('networkidle');
+
+// code will go here
+
+await page.waitForTimeout(10000)
+
+await browser.close();
+</marked-tab>
+<marked-tab header="Puppeteer" lang="javascript">
+import puppeteer from 'puppeteer';
+
+const browser = await puppeteer.launch({ headless: false });
+const page = await browser.newPage();
+
+await page.goto('https://demo-webstore.apify.org/search/on-sale');
+
+await page.waitForNetworkIdle();
+
+// code will go here
+
+await page.waitForTimeout(10000)
+
+await browser.close();
+</marked-tab>
+```
+
+> Notice the slightly different syntax between Playwright and Puppeteer with `waitForLoadState('networkidle')` and `waitForNetworkIdle()`. Sometimes, the differences are fairly subtle like this, but later on we'll run into some more significant differences.
+
+## [](#collecting-in-page-evaluate) Collecting in the browser context
+
+Whatever is returned from the callback function in `page.evaluate()` will be returned by the evaluate function, which means that we can set it to a variable like so:
+
+```JavaScript
+const products = await page.evaluate(() => {});
+```
+
+We'll be returning a bunch of product objects from this function, which will be accessible back in our Node.js context after the promise has resolved. Let's now go ahead and write some data collection code to collect each product:
+
+```JavaScript
+const products = await page.evaluate(() => {
+    return [...document.querySelectorAll('a[class*="ProductCard_root"]')].map((elem) => {
```

Comment:
It's a tutorial. Rather create more variables and use expressive code like `Array.from` than trying to leetcode it. The example becomes squashed and the map is hard to read this way because it's too far right. People might be coming to this from different language backgrounds.
![CleanShot 2022-05-06 at 17 08 13@2x](https://user-images.githubusercontent.com/23726914/167160971-52d6ba55-6577-4058-8ec1-b1f3f1fea707.png)
.